docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+first (e.g.  rebalance), with the removal not taking place until the non-voting
+replica catches up and is upgraded to a full group member.
+
+However, this is listed this as a drawback because going forward with relocation
+*without* non-voting replication introduces this fragility, regardless of how
+relocation decisions are made.
+
+# Alternatives
+The main alternative would be some sort of stateful system, where relocation
+operations would be expressed explicitly (i.e. "Move replica from X to Y") and
+then seen to completion. For reasons outlined in the "motiviation" section, this
+is considered sub-optimal when making distributed decisions.
+
+### Relocation Master
+The ultimate expression of this would be the designation of a "Relocation
+master", a single node that makes all relocation decisions for the entire
+cluster
+
+There is enough information in gossip for a central node to make acceptable
+decisions about replica movement.  In some ways the individual decisions would
+be worse, because they would be unable to consider current raft state; however,
+in aggregate the decisions could be much better, because the central master
+could consider groups of relocations together. For example, it would be able to
+avoid distributed pitfalls such as under-loaded nodes being suddenly inundated
+with requests for new replicas. It would be able to more quickly and correctly
+move replicas to new nodes introduced to the system.
+
+This system would also have an easier time communicating the individual
+relocation decisions that were made. This could be helpful for debugging or
+tweaking relocation criteria.
+
+However, It's important to note that a relocation master is not entirely
+incompatible with the "stateless" design; the relocation master could simply be
+the initiator of the stateless operations. You could thus get much of the
+improved decision making and communication of the master without having to move
+to a stateful design.
+
+# Unresolved questions
+
+### Range leader constraint
+The biggest unresolved issue is the requirement for relocation operations to be
+constrained to replicas with a leader lease.
+
+For example, depending on the criteria there is no firm need for a relocation
+operation to be initiated from a range leader; a secondary replica can safely
+initiate an add or remove replica, with the safety being guaranteed by a
+combination of raft and cockroach's transactions.
+
+However, if the criteria for an operation wants to consider raft state (such as
+which nodes are behind in replication), those decisions could only be made from
+the raft leader. 
+
+Unfortunately, the raft leader is not guaranteed to be the range leader; if we
+require both of these, its possible that no replica on a particular range would
+be able to make a relocation decision. In order to make this work, we would have
+to constrain this operation to the raft leader only, or make a better attempt to
+co-locate the range leader and the raft leader.
```

Comment:
Or introduce some interface by which the raft leader could be queried for this information.